pkg/inotify-informer/inotify.go

```diff
@@ -0,0 +1,177 @@
+}
+
+func splitFileNamespaceName(fullPath string) (namespace string, domain string, err error) {
+	fileName := filepath.Base(fullPath)
+	namespaceName := strings.Split(fileName, "_")
+	if len(namespaceName) != 2 {
+		return "", "", fmt.Errorf("Invalid file path: %s", fullPath)
+	}
+
+	namespace = namespaceName[0]
+	domain = strings.Split(namespaceName[1], ".")[0]
+	return namespace, domain, nil
+}
+
+func (d *DirectoryListWatcher) startBackground() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	var err error
+	if d.backgroundWatcherStarted == true {
+		return nil
+	}
+
+	d.stopChan = make(chan struct{})
+	d.eventChan = make(chan watch.Event)
+
+	d.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	err = d.watcher.Add(d.fileDir)
+	if err != nil {
+		return err
+	}
+
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		for {
+			var e watch.EventType
+			var fse fsnotify.Event
+			select {
+			case <-d.stopChan:
+				d.watcher.Close()
+				return
+			case event := <-d.watcher.Events:
+				fse = event
+				switch event.Op {
+				case fsnotify.Create:
+					e = watch.Added
+				case fsnotify.Remove:
+					e = watch.Deleted
+				}
+			case err := <-d.watcher.Errors:
+				d.eventChan <- watch.Event{
+					Type: watch.Error,
+					Object: &v1.Status{
+						Status: v1.StatusFailure, Message: err.Error(),
+					},
+				}
+			}
+			namespace, name, err := splitFileNamespaceName(fse.Name)
+			if err != nil {
+				logging.DefaultLogger().Error().Reason(err).Msg("Invalid content detected, ignoring and continuing.")
+				continue
+			}
+			d.eventChan <- watch.Event{Type: e, Object: api.NewMinimalDomainWithNS(namespace, name)}
+		}
+	}()
+
+	d.backgroundWatcherStarted = true
+	return nil
+}
+
+func (d *DirectoryListWatcher) List(options v1.ListOptions) (runtime.Object, error) {
+
+	// This starts the watch already.
+	// Starting watching before the actual sync, has the advantage, that we don't
+	// miss notifications about file changes.
+	// It also means that we can't reliably follow file system changes, because we
+	// are informed at least once about changes.
+	err := d.startBackground()
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := ioutil.ReadDir(d.fileDir)
+	if err != nil {
+		d.Stop()
+		return nil, err
+	}
+
+	domainList := &api.DomainList{
+		Items: []api.Domain{},
+	}
+	for _, file := range files {
+		namespace, name, err := splitFileNamespaceName(file.Name())
+		if err != nil {
+			logging.DefaultLogger().Error().Reason(err).Msg("Invalid content detected, ignoring and continuing.")
+			continue
+		}
+		domainList.Items = append(domainList.Items, *api.NewMinimalDomainWithNS(namespace, name))
+
+	}
+	return domainList, nil
+}
+
+func (d *DirectoryListWatcher) Watch(options v1.ListOptions) (watch.Interface, error) {
+	return d, nil
+}
+
+func (d *DirectoryListWatcher) Stop() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.backgroundWatcherStarted == false {
+		return
+	}
+	close(d.stopChan)
+	d.wg.Wait()
```

Comment:
Should the watcher stop asap or really process all events? can we verify that by looking into a k8s listwatcher?

Reply:
The Stop() function for k8s steam watchers closes the connection immediately. For those watchers, only the events in the local queue can be processed after Stop is called. That's pretty close to what the inotify informer is doing.

I do the wg.Wait just to guarantee the goroutine exits before Stop() returns. I didn't like the idea of a start/stop/start potentially briefly involving two overlapping background threads.